### PR TITLE
Add cache for relation reference field formatter

### PR DESCRIPTION
**Pull request by m-kuhn**
this speeds up tables with relation reference fields massively

CC @gioman since you have been suffering from attribute table loading performance issues, maybe this helps?

### CODE REVIEW

**Comment by elpaso** [comment]
@m-kuhn I don't think this helps, I've been debugging @gioman issue and profiling with valgrind, and here is my conclusion: https://github.com/qgis/QGIS/pull/9219#issue-254651574

What would slightly help here is proper caching of widget configuration, however, there is no much gain because the slowdow is really caused by the calls to pg backend to retrieve enum information, which with my patch only happens once (instead of multiple times).

**Comment by m-kuhn** [comment]
This here helps for the same issue but a different widget.

**Comment by elpaso** [comment]
Sure, much appreciated :+1:

File: src/core/fieldformatter/qgsrelationreferencefieldformatter.cpp [hunk 1]
```diff
@@ -90,3 +93,62 @@ QVariant QgsRelationReferenceFieldFormatter::sortValue( QgsVectorLayer *layer, i
...
+  if ( referencedFieldIdx == -1 )
+  {
+    QgsMessageLog::logMessage( QObject::tr( "Invalid referenced field (%1) configured in relation %2" ).arg( relation.fieldPairs().at( 0 ).second, relation.name() ) );
+    return QVariant();
+  }
+
+  QgsFeature feature;
+  auto iterator = referencedLayer->getFeatures();
+
+  QgsExpression expr( referencedLayer->displayExpression() );
```
Comment (nyalldawson) [inline src/core/fieldformatter/qgsrelationreferencefieldformatter.cpp]: Prepare expression

File: src/core/fieldformatter/qgsrelationreferencefieldformatter.cpp [hunk 2]
```diff
@@ -90,3 +93,62 @@ QVariant QgsRelationReferenceFieldFormatter::sortValue( QgsVectorLayer *layer, i
...
+  }
+  int referencedFieldIdx = referencedLayer->fields().lookupField( relation.fieldPairs().at( 0 ).second );
+  if ( referencedFieldIdx == -1 )
+  {
+    QgsMessageLog::logMessage( QObject::tr( "Invalid referenced field (%1) configured in relation %2" ).arg( relation.fieldPairs().at( 0 ).second, relation.name() ) );
+    return QVariant();
+  }
+
+  QgsFeature feature;
+  auto iterator = referencedLayer->getFeatures();
```
Comment (nyalldawson) [inline src/core/fieldformatter/qgsrelationreferencefieldformatter.cpp]: This request could be tightened to only get geometry when needed and only attributes required by the expression.

File: src/core/fieldformatter/qgsrelationreferencefieldformatter.cpp (hunk 1 above)
Comment (m-kuhn) [reply src/core/fieldformatter/qgsrelationreferencefieldformatter.cpp]: :laughing:

**Comment by gioman** [comment]
> this speeds up tables with relation reference fields massively

@m-kuhn I tested @elpaso patch and works fine here. Now that are all merged in master I will test also yours.

**Pull request merged**